Templates.md - proofreading and translating (#5196)
werf/werf: docs/pages_en/usage/deploy/templates.md

{{ $.mykey }}
```

{% endraw %}

Output:

```
myvalue
```

The root context remains intact even in blocks that change the relative context (except for `define`):

{% raw %}

```
{{ with $.Values.backend }}
- command: {{ .command }}
  image: {{ $.Values.werf.image.backend }}
{{ end }}
```

{% endraw %}

Functions like `tpl` or `include` can lose the root context. You can pass the root context as an argument to them to restore access to it:

{% raw %}

```
{{ tpl "{{ .Values.mykey }}" $ }}
```

{% endraw %}

Output:

```
myvalue
```

### Relative context (.)

The relative context is any type of data referenced by the `.` variable. By default, the relative context points to the root context.

Some blocks and functions can modify the relative context. In the example below, in the first line, the relative context points to the root context `$`, while in the second line, it points to `$.Values.containers`:

{% raw %}

```
{{ range .Values.containers }}
{{ . }}
{{ end }}
```

{% endraw %}

Use the `with` block to modify the relative context:

{% raw %}

```
{{ with $.Values.app }}
image: {{ .image }}
{{ end }}
```

{% endraw %}

## Reusing templates

### Named templates

> To reuse templates, declare *named templates* in the `define` blocks in the `templates/_*.tpl` files:

Author: ```suggestion
To reuse templating, declare *named templates* in the `define` blocks in the `templates/_*.tpl` files:
```

{% raw %}

```
# templates/_helpers.tpl:
{{ define "labels" }}
app: myapp
team: alpha
{{ end }}
```

{% endraw %}

Next, insert the named templates into the `templates/*.(yaml|tpl)` files using the `include` function:

{% raw %}

```
# templates/deployment.yaml:
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myapp
spec:
  selector:
    matchLabels: {{ include "labels" nil | nindent 6 }}
  template:
    metadata:
      labels: {{ include "labels" nil | nindent 8 }}
```

{% endraw %}

Output:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myapp
spec:
  selector:
    matchLabels:
      app: myapp
      team: alpha
  template:
    metadata:
      labels:
        app: myapp
        team: alpha
```

The name of the named template to use in the `include` function may be dynamic:

{% raw %}

```
{{ include (printf "%s.labels" $prefix) nil }}
```

{% endraw %}

**Named templates are globally visible** - once declared in a parent or any child chart, a named template becomes available in all charts at once: in both parent and child charts. Make sure there are no named templates with the same name in the parent and child charts.

### Parameterizing named templates

> The `include` function that inserts named templates takes a single optional argument. It can be used to parameterize a named template, where that argument becomes the `.` relative context:

Author: ```suggestion
The `include` function that inserts named templates takes a single optional argument. This argument can be used to parameterize a named template, where that argument becomes the `.` relative context:
```

{% raw %}

```
{{ include "labels" "myapp" }}
```

{% endraw %}

{% raw %}

```
{{ define "labels" }}
app: {{ . }}
{{ end }}
```

{% endraw %}

Output:

```yaml
app: myapp
```

> To pass several arguments at once, use a list containing multiple values:

Author: ```suggestion
To pass several arguments at once, use a list containing multiple arguments:
```

{% raw %}

```
{{ include "labels" (list "myapp") }}
{{ include "labels" (list "myapp" "alpha") }}
```

{% endraw %}

{% raw %}

```
{{ define "labels" }}
app: {{ index . 0 }}
{{ if gt (len .) 1 }}
team: {{ index . 1 }}
{{ end }}
{{ end }}
```

{% endraw %}

Optional non-positional arguments can be handled as follows:

{% raw %}

```
{{ include "labels" (dict "app" "myapp") }}
{{ include "labels" (dict "team" "alpha" "app" "myapp") }}
```

{% endraw %}

{% raw %}

```
{{ define "labels" }}
app: {{ .app }}
{{ if hasKey . "team" }}
team: {{ .team }}
{{ end }}
{{ end }}
```

{% endraw %}

Pass `nil` to a named template that does not require parametrizing:

{% raw %}

```
{{ include "labels" nil }}
```

{% endraw %}

### The result of running include

The `include` function that inserts a named template **returns text data**. To return structured data, you need to *deserialize* the result of `include` using the `fromYaml` function:

{% raw %}

```
{{ define "commonLabels" }}
app: myapp
{{ end }}
```

{% endraw %}

{% raw %}

```
{{ $labels := include "commonLabels" nil | fromYaml }}
{{ $labels.app }}
```

{% endraw %}

Output:

```
myapp
```

> > Note that `fromYaml` does not support lists. Use `fromYamlArray`.

Author: ```suggestion
> Note that `fromYaml` does not support lists. For lists, use the dedicated `fromYamlArray` function.
```

You can use the `toYaml` and `toJson` functions for data serialization, and the `fromYaml/fromYamlArray` and `fromJson/fromJsonArray` functions for deserialization.

### Named template context

The named templates declared in `templates/_*.tpl` cannot use the root and relative contexts of the file into which they are included by the `include` function. You can fix this by passing the root and/or relative context as `include` arguments:

{% raw %}

```
{{ include "labels" $ }}
{{ include "labels" . }}
{{ include "labels" (list $ .) }}
{{ include "labels" (list $ . "myapp") }}
```

{% endraw %}

### include in include

You can also use the `include` function in the `define` blocks to include named templates:

{% raw %}

```
{{ define "doSomething" }}
{{ include "doSomethingElse" . }}
{{ end }}
```

{% endraw %}

> You can even use  the `include` function to include the named template recursively:

Author: ```suggestion
You can even call the `include` function to include a named template from this very template, i.e., recursively:
```
